compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4547,10 +4580,79 @@ public void visit(BLangUnaryExpr unaryExpr) {
+            if (OperatorKind.SUB.equals(unaryExpr.operator)) {
+                LinkedHashSet<BType> basicNumericTypes = new LinkedHashSet<>();
+                BType referredType = types.getReferredType(expType);
+
+                if (expType == symTable.noType) {
+                    newExpectedType = symTable.noType;
+                } else if (TypeTags.isIntegerTypeTag(referredType.tag) || referredType.tag == TypeTags.FLOAT
+                        || referredType.tag == TypeTags.DECIMAL) {
+                    newExpectedType = referredType;
+                } else if (referredType.tag == TypeTags.FINITE) {
+                    BFiniteType finiteType = (BFiniteType) referredType;
+                    for (BLangExpression value : finiteType.getValueSpace()) {
+                        int typeTag = value.getBType().tag;
+                        if (TypeTags.isIntegerTypeTag(typeTag) || typeTag == TypeTags.FLOAT
+                                || typeTag == TypeTags.DECIMAL) {
+                            basicNumericTypes.add(value.getBType());
+                        } else if (typeTag == TypeTags.JSON || typeTag == TypeTags.ANYDATA || typeTag == TypeTags.ANY) {
+                            basicNumericTypes.add(symTable.intType);
+                            basicNumericTypes.add(symTable.floatType);
+                            basicNumericTypes.add(symTable.decimalType);
+                            break;
+                        }
+                    }
+                    if (basicNumericTypes.isEmpty()) {
+                        dlog.error(unaryExpr.pos, DiagnosticErrorCode.UNARY_OP_INCOMPATIBLE_TYPES,
+                                unaryExpr.operator, unaryExpr.expr);
+                        resultType = symTable.semanticError;
+                    } else if (basicNumericTypes.size() == 1) {
+                        newExpectedType = basicNumericTypes.iterator().next();
+                    } else {
+                        newExpectedType = BUnionType.create(null, basicNumericTypes);
+                    }
+                } else if (referredType.tag == TypeTags.UNION) {
+                    BUnionType referredUnionType = (BUnionType) referredType;
+                    for (BType value : referredUnionType.getMemberTypes()) {
+                        int typeTag = value.tag;
+                        if (TypeTags.isIntegerTypeTag(typeTag) || typeTag == TypeTags.FLOAT
+                                || typeTag == TypeTags.DECIMAL) {
+                            basicNumericTypes.add(value);
+                        } else if (typeTag == TypeTags.JSON || typeTag == TypeTags.ANYDATA || typeTag == TypeTags.ANY) {
+                            basicNumericTypes.add(symTable.intType);
+                            basicNumericTypes.add(symTable.floatType);
+                            basicNumericTypes.add(symTable.decimalType);
+                            break;
+                        }
```

Comment:
If we extract out the relevant numeric type identification logic to a separate method, we can recursively call that here.